polars/polars-core/src/chunked_array/ops/explode.rs

```diff
@@ -304,48 +304,16 @@ impl ExplodeByOffsets for BinaryChunked {
 
 /// Convert Arrow array offsets to indexes of the original list
 pub(crate) fn offsets_to_indexes(offsets: &[i64], capacity: usize) -> Vec<IdxSize> {
-    if offsets.is_empty() {
-        return vec![];
-    }
-    let mut idx = Vec::with_capacity(capacity);
-
-    // `value_count` counts the taken values from the list values
-    // and are the same unit as `offsets`
-    // we also add the start offset as a list can be sliced
-    let mut value_count = offsets[0];
-    // `empty_count` counts the duplicates taken because of empty list
-    let mut empty_count = 0usize;
-    let mut last_idx = 0;
-
-    for offset in &offsets[1..] {
-        // this get all the elements up till offsets
-        while value_count < *offset {
-            value_count += 1;
-            idx.push(last_idx)
-        }
-
-        // then we compute the previous offsets
-        // Safety:
-        // we started iterating from 1, so there is always a previous offset
-        // we take the pointer to the previous element and deref that to get
-        // the previous offset
-        let previous_offset = unsafe { *(offset as *const i64).offset(-1) };
-
-        // if the previous offset is equal to the current offset we have an empty
-        // list and we duplicate previous index
-        if previous_offset == *offset {
-            empty_count += 1;
-            idx.push(last_idx);
-        }
-
-        last_idx += 1;
-    }
-
-    // take the remaining values
-    for _ in 0..(capacity - (value_count - offsets[0]) as usize - empty_count) {
-        idx.push(last_idx);
-    }
-    idx
+    offsets
```

Comment:
Thanks for making a PR.

Though I don't doubt this solves the issue, on first sight this looks much more expensive thab the previous implementation.

A chained iterator is much slower, and vec allocations in a loop as well. I'd like a fix that doesn't regress the performance of this code as we use this a lot.

Reply:
That's a fair point, I didn't really test the performance here. I'll try to look for a faster implementation. Is there a particular benchmark I could use to test the performance of this function? If not, I can set one up.

Reply:
Nope, just create some data with over a million real world like offset. (A million to ensure we are not in cache)

And maybe also benchmark on 10k data points as that may fit in cache.

Reply:
Got it, will do. Thanks for the feedback!

Reply:
You were right about the timing; my implementation with zip was about 2.5 times slower across all input sizes I tried. I tried inputs with capacities 10, 10000, and 10000000, with various numbers of offsets between 0 and `capacity`. I can provide more details and/or my benchmarking script if you want.

One simple alternative to my zip implementation would be to use the original code but replace this portion at the end:
```
// take the remaining values
for _ in 0..capacity - (value_count - offsets[0]) as usize - empty_count {
    idx.push(last_idx);
}
idx
```

with this code:
```
// take the remaining values
for _ in 0..capacity.saturating_sub(idx.len()) {
    idx.push(last_idx);
}
if idx.len() > capacity {
    idx.into_iter().take(capacity).collect()
} else {
    idx
}
```

In the case that the original implementation doesn't underflow, the timing is basically identical because the same amount of values are pushed as before and the saturating_sub has a negligible overhead. In the case where there was an underflow before, the new code copies the subset of the vector required to keep the length equal to capacity, which adds an O(capacity) overhead. However, since this case would run infinitely in release before, I think any finite-time performance is an improvement. Because the new code also doesn't use the empty_count value, the empty_count increments can be removed from the main loop, though this doesn't seem to actually provide a speedup.

A slightly better approach might be to alter the main loop of the original implementation to prevent copying more indices than the capacity in the first place. However, I worry that would add more overhead in the typical case and make the implementation more complex overall. I'll keep looking at alternatives, unless you think the saturating_sub option is sufficient.

Reply:
Your second suggestion is almost there.

```
// take the remaining values
for _ in 0..capacity.saturating_sub(idx.len()) {
    idx.push(last_idx);
}
if idx.len() > capacity {
    idx.into_iter().take(capacity).collect()
} else {
    idx
}

```

If instead you'd do

```
// take the remaining values
for _ in 0..capacity.saturating_sub(idx.len()) {
    idx.push(last_idx);
}
idx.truncate(capacity)
```

It would not re allocate